### PR TITLE
changed: use libfstrcmp

### PR DESCRIPTION
**Pull request by notspiff**
Move to libfstrmp instead of fstrcmp.h/c - this is a GPL implementation.

Still needs some work (upload of source tar ball, windows changes).

the bsd build is a total guess work.

### CODE REVIEW

File: cmake/modules/Findfstrcmp.cmake [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+# Findfstrcmp
+# --------
+# Finds the fstrcmp library
+#
+# This will define the following variables::
+#
+# FSTRCMP_FOUND - system has libfstrcmp
+# FSTRCMP_INCLUDE_DIRS - the libfstrcmp include directory
+# FSTRCMP_LIBRARIES - the libfstrcmp libraries
+# FSTRCMP_DEFINITIONS - the libfstrcmp compile definitions
```
Comment (Rechi) [inline cmake/modules/Findfstrcmp.cmake]: not set

File: tools/depends/target/libfstrcmp/Makefile [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+# lib name, version
+LIBNAME=fstrcmp
+VERSION=0.7.D001
+SOURCE=$(LIBNAME)-$(VERSION)
+ARCHIVE=$(SOURCE).tar.gz
+
+# configuration settings
+CONFIGURE=./configure --prefix=$(PREFIX) --disable-shared
+
+LIBDYLIB=$(PLATFORM)/.libs/$(LIBNAME)16.a
```
Comment (Rechi) [inline tools/depends/target/libfstrcmp/Makefile]: this looks like it is copied from libpng

File: xbmc/Util.cpp [hunk 3]
```diff
@@ -62,7 +62,7 @@
 #include "utils/RegExp.h"
 #include "windowing/GraphicContext.h"
 #include "guilib/TextureManager.h"
-#include "utils/fstrcmp.h"
+#include "fstrcmp.h"
```
Comment (Rechi) [inline xbmc/Util.cpp]: `<fstrcmp.h>` and include order

File: xbmc/addons/Scraper.cpp [hunk 4]
```diff
@@ -41,7 +41,7 @@
 #include "utils/StringUtils.h"
 #include "utils/URIUtils.h"
 #include "utils/XMLUtils.h"
-#include "utils/fstrcmp.h"
+#include <fstrcmp.h>
```
Comment (Rechi) [inline xbmc/addons/Scraper.cpp]: include order

File: xbmc/utils/StringUtils.cpp [hunk 5]
```diff
@@ -37,7 +37,7 @@
 #include "StringUtils.h"
 #include "CharsetConverter.h"
 #include "LangInfo.h"
-#include "utils/fstrcmp.h"
+#include <fstrcmp.h>
```
Comment (Rechi) [inline xbmc/utils/StringUtils.cpp]: include order

File: tools/buildsteps/freebsd/make-depends [hunk 6]
```diff
@@ -2,4 +2,7 @@ WORKSPACE=${WORKSPACE:-$( cd $(dirname $0)/../../.. ; pwd -P )}
 XBMC_PLATFORM_DIR=freebsd
 . $WORKSPACE/tools/buildsteps/defaultenv
 
-#nothing for freebsd atm
+if [ "$(pathChanged $WORKSPACE/tools/depends)" == "1" ]
+then
+  cd $WORKSPACE/tools/depends;make -C target/libfstrcmp -j $BUILDTHREADS && tagSuccessFulBuild $WORKSPACE/tools/depends
```
Comment (Rechi) [inline tools/buildsteps/freebsd/make-depends]: that doesn't work

**Review by Rechi — CHANGES_REQUESTED** [review]

File: tools/depends/target/libfstrcmp/Makefile (hunk 2 above)
Comment (notspiff) [reply tools/depends/target/libfstrcmp/Makefile]: aye.

File: tools/buildsteps/freebsd/make-depends (hunk 6 above)
Comment (notspiff) [reply tools/buildsteps/freebsd/make-depends]: i suspected as much, i didn't have anything to monkey.

File: xbmc/Util.cpp [hunk 7]
```diff
@@ -58,11 +58,11 @@
 #ifdef HAS_UPNP
 #include "filesystem/UPnPDirectory.h"
 #endif
+#include <fstrcmp.h>
```
[the remark this answers is not in the document]
Comment (notspiff) [reply xbmc/Util.cpp]: as the file doesn't follow the guide lines it's hard to see what's correct. i did another attempt.

[the remark this answers is not in the document]
Comment (Rechi) [reply xbmc/Util.cpp]: better now, can you fix the others too

File: tools/buildsteps/freebsd/make-depends [hunk 8]
```diff
@@ -2,4 +2,4 @@ WORKSPACE=${WORKSPACE:-$( cd $(dirname $0)/../../.. ; pwd -P )}
 XBMC_PLATFORM_DIR=freebsd
 . $WORKSPACE/tools/buildsteps/defaultenv
 
-#nothing for freebsd atm
+# nothing for freebsd atm
```
Comment (Rechi) [inline tools/buildsteps/freebsd/make-depends]: unrelated

File: CMakeLists.txt [hunk 9]
```diff
@@ -118,6 +119,7 @@ set(required_deps ASS
                   Fmt
                   FreeType
                   FriBidi
+                  fstrcmp
```
Comment (wsnipex) [inline CMakeLists.txt]: this also need to be added to the kodi binary depends on L287

File: tools/depends/target/libfstrcmp/Makefile [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+$(TARBALLS_LOCATION)/$(ARCHIVE):
+	cd $(TARBALLS_LOCATION); $(RETRIEVE_TOOL) $(RETRIEVE_TOOL_FLAGS) $(BASE_URL)/$(ARCHIVE)
+
+$(PLATFORM): $(TARBALLS_LOCATION)/$(ARCHIVE) $(DEPS)
+	rm -rf $(PLATFORM)/*; mkdir -p $(PLATFORM)
+	cd $(PLATFORM); $(ARCHIVE_TOOL) $(ARCHIVE_TOOL_FLAGS) $(TARBALLS_LOCATION)/$(ARCHIVE)
+	cd $(PLATFORM); autoreconf -vif; $(CONFIGURE)
+
+$(LIBDYLIB): $(PLATFORM)
+	$(MAKE) -C $(PLATFORM)
```
Comment (Rechi) [inline tools/depends/target/libfstrcmp/Makefile]:
```
$(MAKE) -j1 -C $(PLATFORM)
touch $@
```

File: tools/depends/target/libfstrcmp/Makefile [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+.installed-$(PLATFORM): $(LIBDYLIB)
+	$(MAKE) -C $(PLATFORM) install
```
Comment (Rechi) [inline tools/depends/target/libfstrcmp/Makefile]:
```
mkdir -p $(PREFIX)/lib/pkgconfig
$(MAKE) -C $(PLATFORM) install-include install-libdir
```

File: tools/depends/target/libfstrcmp/Makefile [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+ifeq ($(OS), $(filter $(OS),osx ios))
+	rm -f $(PREFIX)/lib/libfstrcmp.dylib*
+else
+	rm -f $(PREFIX)/lib/libfstrcmp.so*
+endif
```
Comment (Rechi) [inline tools/depends/target/libfstrcmp/Makefile]: `rm -f $(PREFIX)/lib/libfstrcmp.so* rm -f $(PREFIX)/lib/libfstrcmp.dylib*`

File: cmake/modules/Findfstrcmp.cmake [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+  string(REGEX MATCH "VERSION=[^ ]*" FSTRCMP_VER "${VER}")
+  list(GET FSTRCMP_VER 0 FSTRCMP_VER)
+  string(SUBSTRING "${FSTRCMP_VER}" 8 -1 FSTRCMP_VER)
+
+  # allow user to override the download URL with a local tarball
+  # needed for offline build envs
+  if(FSTRCMP_URL)
+    get_filename_component(FSTRCMP_URL "${FSTRCMP_URL}" ABSOLUTE)
+  else()
+    set(FSTRCMP_URL http://fstrcmp.sourceforge.net/fstrcmp-0.7.D001.tar.gz) # http://mirrors.kodi.tv/build-deps/sources/fstrcmp-${FSTRCMP_VER}.tar.gz)
```
Comment (Rechi) [inline cmake/modules/Findfstrcmp.cmake]: uploaded

**Comment by notspiff** [comment]
jenkins build this please

**Comment by notspiff** [comment]
jenkins build this please

**Comment by Rechi** [comment]
README.FreeBSD.md needs some updated on how to get libfstrcmp or use the internal version.

**Review by Rechi — APPROVED** [review]

**Comment by notspiff** [comment]
jenkins build this please

File: tools/depends/target/libfstrcmp/Makefile [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+	cd $(TARBALLS_LOCATION); $(RETRIEVE_TOOL) $(RETRIEVE_TOOL_FLAGS) $(BASE_URL)/$(ARCHIVE)
+
+$(PLATFORM): $(TARBALLS_LOCATION)/$(ARCHIVE) $(DEPS)
+	rm -rf $(PLATFORM)/*; mkdir -p $(PLATFORM)
+	cd $(PLATFORM); $(ARCHIVE_TOOL) $(ARCHIVE_TOOL_FLAGS) $(TARBALLS_LOCATION)/$(ARCHIVE)
+	cd $(PLATFORM); autoreconf -vif; $(CONFIGURE)
+
+$(LIBDYLIB): $(PLATFORM)
+	$(MAKE) -j1 -C $(PLATFORM) lib/libfstrcmp.la
+	touch $@
```
[the remark this answers is not in the document]
Comment (Rechi) [reply tools/depends/target/libfstrcmp/Makefile]: It was initially required as the all target did something wrong with the file times (rebuild detection). With `lib/libfstrcmp.la` it isn't needed anymore.
Btw, using `lib/libfstrcmp.la` target also fixed the concurrent issue, so `-j1` also is no longer required.

**Comment by notspiff** [comment]
failure is unrelated, all comments addressed. somebody hit the button unless there is more to do please.

**Review by wsnipex — APPROVED** [review]

**Comment by Rechi** [comment]
Which failure? It is still building.

**Comment by notspiff** [comment]
yes, a recycle due to a failed test on osx64.

File: docs/README.openSUSE.md [hunk 15]
```diff
@@ -74,7 +74,7 @@ If you get a `package not found` type of message with the below command, remove
...
+sudo zypper install alsa-devel autoconf automake bluez-devel boost-devel capi4linux-devel ccache cmake doxygen flac-devel fribidi-devel fstrcmp-devel gcc gcc-c++ gettext-devel giflib-devel glew-devel gperf java-openjdk libass-devel libavahi-devel libbluray-devel libbz2-devel libcap-devel libcap-ng-devel libcdio-devel libcec-devel libcurl-devel libdvdread-devel libgudev-1_0-devel libidn2-devel libjasper-devel libjpeg-devel liblcms2-devel libmad-devel libmicrohttpd-devel libmodplug-devel libmpeg2-devel libmysqlclient-devel libnfs-devel libogg-devel libpcap-devel libplist-devel libpng12-devel libpulse-devel libsamplerate-devel libsmbclient-devel libtag-devel libtiff-devel libtool libudev-devel libuuid-devel libva-devel libvdpau-devel libvorbis-devel libXrandr-devel libXrender-devel libxslt-devel libyajl-devel lirc-devel lzo-devel make Mesa-libEGL-devel Mesa-libGLESv2-devel Mesa-libGLESv3-devel nasm patch pcre-devel python-devel randrproto-devel renderproto-devel shairplay-devel sqlite3-devel swig tinyxml-devel
```
[the remark this answers is not in the document]
Comment (candrews) [reply docs/README.openSUSE.md]: FYI, I had to add dev-libs/libfstrcmp to Gentoo just now as fstrcmp wasn't packaged there, either: https://github.com/gentoo/gentoo/commit/2c222f7234ddbd2a73917c06e2b6e44a939e9c8f
